### PR TITLE
Fix a number of memory leaks in validation layers

### PR DESCRIPTION
**Pull request by polarina**
These commits address the leaks reported in #1787 and some more.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -5305,6 +5305,7 @@ static void PostCallRecordDestroyPipeline(layer_data *dev_data, VkPipeline pipel
                                           VK_OBJECT obj_struct) {
     // Any bound cmd buffers are now invalid
     invalidateCommandBuffers(dev_data, pipeline_state->cb_bindings, obj_struct);
+    delete getPipelineState(dev_data, pipeline);
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: Cases like this we should really be using unique_ptr or holding the state by value, but fixing the memory leak is a good step.

**Comment by chrisforbes** [comment]
Bad news, this crashes everywhere but nvidia/linux and mesa-skylake/linux.

**Comment by polarina** [comment]
@chrisforbes What an unfortunate coincidence that those two platforms are the ones I tested on. I'll investigate and get back to you. Thanks for testing.

**Comment by polarina** [comment]
@chrisforbes I see that there's an `gh1841` branch with these commits. I assume the tests were ran against a build from that branch? If so, can you smoke test them with #1842 merged on top? That pull request fixes a heap-use-after-free bug in the tests that the changes in 2fd398e0b0030bada3b021c7ab4f048f1631a205 trigger.

**Comment by chrisforbes** [comment]
@polarina Yes, that branch was what I ran. I will test your new change now.

**Comment by chrisforbes** [comment]
@polarina This PR rebased onto master (which includes #1842) still segfaults in `VkLayerTest.SparseResidencyImageCreateUnsupportedTypes` on Android.

**Comment by chrisforbes** [comment]
@polarina This crash now occurs on the skylake/mesa platform as well, which should make your debugging efforts easier.

**Comment by chrisforbes** [comment]
@polarina some valgrind fixes on `gh1841b`; they don't help with this, however. Disabling those two tests gets us through complete test runs on all targets, so should help narrow this down.

**Comment by polarina** [comment]
@chrisforbes I have found and fixed the cause of these crashes, I hope.

GetLayerDataPtr stores the key and a pointer to newly allocated memory in an unordered_map, `layer_data_map`, and returns that allocation on later calls that give the same `data_key`. Only /some/ of the `vkDestroyInstance` and `vkDestroyDevice` functions in the layers erase the key from `layer_data_map` (and some even attempt to erase from the wrong map).

Since we now `delete` these allocations, later calls to `new` may return a pointer that has been deleted but still lingers on in the `layer_data_map`. When that happens, very bad things happen from there on.

Can you test again, please?

**Comment by chrisforbes** [comment]
Much better :+1:

**Pull request merged**